### DISCUSSION TITLE
Send didSave LSP notification iwth Monaco Language Client

### DISCUSSION THREAD

**Discussion by theogiraudet** [TypeFox/monaco-languageclient#348] [Q&A] [answered]
Hello,
I search a way to send a didSave notification to the server when the user do CTRL+S for example (or click on a button).
Is there a way to do it with Monaco Language Client? I've looked at the code a bit and found some pieces of code corresponding to this feature but I haven't found how to use it.

Thanks for your help!

**Comment by CGNonofr** [comment, marked as answer]
Hello,

You need to have your own implementation of `Workspace`, you can extends `MonacoWorkspace` and define `onWillSaveTextDocument` and `onDidSaveTextDocument`

You'll need to use `Services.install` instead of `MonacoServices.install` to be able to provide your custom implementation (take a look at `monaco-services.ts`)

**Reply by theogiraudet** [reply]
Thanks you for your answer!
I have made my own implementations of `MonacoWorkspace` and `Services`. I have getter to access to `onDidSaveTextDocument` event but `Event` is only listenable. The only way I can see to trigger it (when the user do CTRL+S for example) is to directly do it on the emitter. However, I don't see how to access to the current TextDocument to pass to my emitter.
In MonacoWorkspace, I see that to call `onDidOpenTextDocumentEmitter` for example, the fire event is bound to the `onDidCreateModel` event from `monaco.editor`. But an equivalent event doesn't exist on this class for the saving.
So, how can I call properly the `onDidSaveTextDocument` event?

Thanks!

<details>
  <summary>My code</summary>

My `CustomWorkspace`:

```typescript
import { MonacoToProtocolConverter, ProtocolToMonacoConverter, Emitter, MonacoWorkspace, 
    TextDocument, TextDocumentWillSaveEvent } from '@codingame/monaco-languageclient';
import { Event } from "@codingame/monaco-languageclient"
import type * as monaco from 'monaco-editor-core';

export class CustomWorkspace extends MonacoWorkspace {

    protected readonly onDidSaveTextDocumentEmitter = new Emitter<TextDocument>();
    protected readonly onWillSaveTextDocumentEmitter = new Emitter<TextDocumentWillSaveEvent>();

    constructor(
        protected readonly _monaco: typeof monaco,
        protected readonly p2m: ProtocolToMonacoConverter,
        protected readonly m2p: MonacoToProtocolConverter,
        protected _rootUri: string | null = null
    ) {
        super(_monaco, p2m, m2p, _rootUri)
    }

    get onDidSaveTextDocument(): Event<TextDocument> {
        return this.onDidSaveTextDocumentEmitter.event;
    }

    get onWillSaveTextDocument(): Event<TextDocumentWillSaveEvent> {
        return this.onWillSaveTextDocumentEmitter.event;
    }
}
```

My `CustomServices`:
```typescript
import { ConsoleWindow } from "@codingame/monaco-languageclient/"
import { MonacoCommands } from "@codingame/monaco-languageclient/"
import { MonacoToProtocolConverter, ProtocolToMonacoConverter } from "@codingame/monaco-languageclient"
import { MonacoLanguages } from "@codingame/monaco-languageclient/"
import { Services } from "@codingame/monaco-languageclient/"
import type * as monaco from 'monaco-editor-core';
import { CustomWorkspace } from './CustomWorkspace'

export interface CustomServices extends Services {
    commands: MonacoCommands
    languages: MonacoLanguages
    workspace: CustomWorkspace
    window: ConsoleWindow
}

export namespace CustomServices {
    export interface Options {
        rootUri?: string
    }
    export type Provider = () => CustomServices;
    export function create(_monaco: typeof monaco, options: Options = {}): CustomServices {
        const m2p = new MonacoToProtocolConverter(_monaco);
        const p2m = new ProtocolToMonacoConverter(_monaco);
        return {
            commands: new MonacoCommands(_monaco),
            languages: new MonacoLanguages(_monaco, p2m, m2p),
            workspace: new CustomWorkspace(_monaco, p2m, m2p, options.rootUri),
            window: new ConsoleWindow()
        }
    }
    export function install(_monaco: typeof monaco, options: Options = {}): CustomServices {
        const services = create(_monaco, options);
        Services.install(services);
        return services;
    }
    export function get(): CustomServices {
        return Services.get() as CustomServices;
    }
}
```

The current way I use to execute action at CTRL+S:
```typescript
function bindSaveHotkey(client: MonacoLanguageClient, services: CustomServices) {
    document.getElementById('container')!.addEventListener("keydown", (e) => {
      if (e.code === "KeyS" && (navigator.platform.match("Mac") ? e.metaKey : e.ctrlKey)) {
        e.preventDefault();
        console.log("Saved")
      }
    }, false);
  }
```
</details>

**Reply by CGNonofr** [reply]
You have to implement it by hands

`CustomWorkspace`:
```
export class CustomWorkspace extends MonacoWorkspace {

    protected readonly onDidSaveTextDocumentEmitter = new Emitter<TextDocument>();
    protected readonly onWillSaveTextDocumentEmitter = new Emitter<TextDocumentWillSaveEvent>();

    constructor(
        protected readonly _monaco: typeof monaco,
        protected readonly p2m: ProtocolToMonacoConverter,
        protected readonly m2p: MonacoToProtocolConverter,
        protected _rootUri: string | null = null
    ) {
        super(_monaco, p2m, m2p, _rootUri)
    }

    get onDidSaveTextDocument(): Event<TextDocument> {
        return this.onDidSaveTextDocumentEmitter.event;
    }

    get onWillSaveTextDocument(): Event<TextDocumentWillSaveEvent> {
        return this.onWillSaveTextDocumentEmitter.event;
    }

   saveDocument (document: TextDocument, reason: TextDocumentSaveReason) {
      this.onWillSaveTextDocumentEmitter.fire({
          textDocument: document,
          reason
      })

     [save the file here]

    this.onDidSaveTextDocumentEmitter.fire(document)
  }
}
```

Then you can call `(Services.get() as CustomServices).workspace.save(TextDocument.create(....), reason)`

**Reply by theogiraudet** [reply]
Thanks you, now when I do the CTRL+S the LSP notification is sent!
However any action just after (Hover, typing/removing character...) makes LSP server crash with "Header must provide a Content-Length property" error. I will search why this bug occurs.

**Reply by CGNonofr** [reply]
Just a blind guess, do you use bluebird in your backend?

**Reply by theogiraudet** [reply]
Hmm... No I don't think, my backend is at this moment the Langium Hello World with a listener on the didSave receive event. The error is thrown by `\vscode-jsonrpc\lib\common\messageReader`.

**Reply by theogiraudet** [reply]
But `bfj` is used in my frontend, it's a transitive dependency from `react-script`.

**Reply by CGNonofr** [reply]
I had this error several times, because of [this condition](https://github.com/microsoft/vscode-languageserver-node/blob/d58c00bbf8837b9fd0144924db5e7b1c543d839e/jsonrpc/src/common/semaphore.ts#L62) returning false, due to the Promise object being overwritten by a lib like bluebird

**Reply by theogiraudet** [reply]
Oh I see, do you think it can be the same on frontend?

**Reply by CGNonofr** [reply]
If you use a websocket, this error can only come from the backend in the code transforming websocket messages <=> process stream messages

**Reply by theogiraudet** [reply]
Okay, I'll check that out. thanks for your help, I'll mark the discussion as answered.
